node.go

```diff
@@ -0,0 +1,325 @@
+package amt
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/filecoin-project/go-amt-ipld/v3/internal"
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+type node struct {
+	// these may both be nil if the node is empty (a root node)
+	links  [internal.Width]*link
+	values [internal.Width]*cbg.Deferred
+}
+
+var (
+	errEmptyNode      = errors.New("unexpected empty amt node")
+	errUndefinedCID   = errors.New("amt node has undefined CID")
+	errLinksAndValues = errors.New("amt node has both links and values")
+	errLeafUnexpected = errors.New("amt leaf not expected at height")
+	errLeafExpected   = errors.New("amt expected at height")
+	errInvalidCount   = errors.New("amt count does not match number of elements")
+)
+
+func newNode(nd internal.Node, allowEmpty, expectLeaf bool) (*node, error) {
+	if len(nd.Links) > 0 && len(nd.Values) > 0 {
+		return nil, errLinksAndValues
+	}
+
+	i := 0
+	n := new(node)
+	if len(nd.Values) > 0 {
+		if !expectLeaf {
+			return nil, errLeafUnexpected
+		}
+		for x := byte(0); x < internal.Width; x++ {
```

Comment:
nit: can we drop the cast to byte to help avoid confusion? Reading through the dense bit operations on line 42 I kept getting thrown by byte x not having bit ops performed on it. I'm missing how the cast helps since the loop bounds should keep all array indexes in bounds.

Reply:
I believe this was a micro optimization but it almost certainly doesn't make a difference. I'll remove it.